src/Functions/h3PolygonToCells.cpp

```diff
@@ -0,0 +1,187 @@
+#include "config.h"
+
+#if USE_H3
+
+#include <Columns/ColumnArray.h>
+
+#include <Functions/FunctionFactory.h>
+#include <Functions/geometryConverters.h>
+#include <Functions/IFunction.h>
+
+#include <boost/geometry.hpp>
+#include <boost/geometry/geometries/point_xy.hpp>
+
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypesNumber.h>
+
+#include <constants.h>
+#include <h3api.h>
+
+
+// TODO: from h3ToChildren, why this exact number?
+static constexpr size_t MAX_ARRAY_SIZE = 1 << 30;
+
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int ARGUMENT_OUT_OF_BOUND;
+    extern const int TOO_LARGE_ARRAY_SIZE;
+    extern const int ILLEGAL_COLUMN;
+}
+
+class FunctionH3PolygonToCells : public IFunction
+{
+public:
+    static constexpr auto name = "h3PolygonToCells";
+    String getName() const override { return name; }
+    static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionH3PolygonToCells>(); }
+
+    size_t getNumberOfArguments() const override { return 2; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+    // TODO: Is it suitable?
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes &) const override
+    {
+        return std::make_shared<DataTypeArray>(std::make_shared<DataTypeUInt64>());
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        // TODO: what is non-const in this context?
+        //       and why are h3 functions using non-const
+        auto non_const_arguments = arguments;
+        for (auto & argument : non_const_arguments)
+            argument.column = argument.column->convertToFullColumnIfConst();
+
+        const auto * col_resolution = checkAndGetColumn<ColumnUInt8>(non_const_arguments[1].column.get());
+        if (!col_resolution)
+            throw Exception(
+                ErrorCodes::ILLEGAL_COLUMN,
+                "Illegal type {} of argument {} of function {}. Must be UInt8.",
+                arguments[1].type->getName(),
+                2,
+                getName());
+        const auto & data_resolution = col_resolution->getData();
+
+
+        auto dst = ColumnArray::create(ColumnUInt64::create());
+        auto & dst_data = dst->getData();
+        auto & dst_offsets = dst->getOffsets();
+        dst_offsets.resize(input_rows_count);
+        auto current_offset = 0;
+
+
+        callOnGeometryDataType<CartesianPoint>(arguments[0].type, [&] (const auto & type)
```

Comment:
What about spherical coordinates here?

Reply:
I couldn't figure out how to support both, so I went with Cartesian only (which I believe matches some of the [existing h3 functions](https://github.com/ClickHouse/ClickHouse/blob/30991f29ac0148855dd6a926fe1d97afdce87a3d/src/Functions/geoToH3.cpp#L132)). Would it be okay to go ahead without spherical for now?